### ISSUE TITLE
[C+ Checklist Needs Completion] [$1000] Web - Task title input is not focused after reloading the page

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22850]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Go to assign task from FAB
3. Reload the page

## Expected Result:
 Input should remain focused

## Actual Result:
 Input is not focused after reload

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.40-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/559b6b55-2bca-4bd3-8fd1-e94e7223ab42

https://github.com/Expensify/App/assets/93399543/de7e3a4f-7f10-43f9-8c10-96111167df0f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1689271743989309

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e4d7c9a31e0cc909</li>
        <li>Upwork Job ID: 1680028911628701696</li>
        <li>Last Price Increase: 2023-08-05</li>
    </ul>
</details>

**Comment by greg-schroeder** [comment]
Hmm - should I mark this `External` or wait for the outcome of that discussion @bernhardoj? Sounds like we may not know exactly where we want to go with this.

**Comment by gadhiyamanan** [comment]
@greg-schroeder this is dupe of https://github.com/Expensify/App/issues/20000

**Comment by greg-schroeder** [comment]
I'n not going to reopen that issue as this one has a more robust discussion about the issue and proposals already working. I'll list you as the issue reporter, though.

**Comment by zaitsev1** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The first form field should be in focus after a reload

### What is the root cause of that problem?
The `NewTaskDetailsPage.js` form is missing prop autoFocus for the first field

### What changes do you think we should make in order to solve the problem?
Add `autoFocus` prop here:

https://github.com/Expensify/App/blob/2454ea88b15ec4bb6216a2b61acb294b9573fb20/src/pages/tasks/NewTaskDetailsPage.js#L94-L102

the input field will receive focus and the keyboard will be automatically displayed when the component renders.


https://github.com/Expensify/App/assets/32131367/d03e51e1-81df-40e7-a79d-7e5e1c4a945c

**Comment by greg-schroeder** [comment]
Awaiting proposal review from @aimane-chnaif :)

**Comment by zaitsev1** [comment]
## Proposal to @aimane-chnaif [comment](https://github.com/Expensify/App/issues/22850#issuecomment-1640119038)

### Please re-state the problem that we are trying to solve in this issue.
The first form field should be in focus after a reload

### What is the root cause of that problem?
The transitionEnd event is only triggered after certain types of transitions, like when navigating between screens. It does not get triggered on all rendering updates. So, if you are just refreshing the same screen, the event will not fire.

### What changes do you think we should make in order to solve the problem?
Adding a 'ready' event to React Navigation is another useful callback that, in addition to 'transitionEnd', can help ensure that logic is executed on every navigation.

https://github.com/Expensify/App/blob/16a83ac373c2103add1e416d2b33b2365ab1b87f/src/components/ScreenWrapper/index.js#L37-L45

Add below:

```javascript
        this.unsubscribeReady = this.props.navigation.addListener('ready', () => {
            // Runs on each navigation and re-render
            if (!this.props.onEntryTransitionEnd) return;

            this.props.onEntryTransitionEnd();
        });
``` 

Then we need to Remove listener on unmount

https://github.com/Expensify/App/blob/16a83ac373c2103add1e416d2b33b2365ab1b87f/src/components/ScreenWrapper/index.js#L71-L74

Add below:
```javascript
if (this.unsubscribeReady) {
    this.unsubscribeReady();
}
``` 


https://github.com/Expensify/App/assets/32131367/d9cb980c-0f39-4a43-9fdc-135ab92ba381


### What alternative solutions did you explore? (Optional)

in the Form Comopnent find the first input field and assign focus to it

https://github.com/Expensify/App/blob/2454ea88b15ec4bb6216a2b61acb294b9573fb20/src/components/Form.js#L149-L160

```javascript
if (isFirstRender.current) {
    //  Focus the first input when the component mounts
    const firstInputRef = inputRefs.current[_.keys(inputRefs.current)[0]];
    if (firstInputRef && firstInputRef.focus && typeof firstInputRef.focus === 'function') {
        firstInputRef.focus();
    }

    isFirstRender.current = false;
    return;
}
``` 

This will make it always focus the first input, we can also make prop for the condition whether to focus or not.  

https://github.com/Expensify/App/assets/32131367/e4040abe-6674-4e2e-9ed0-236cbc3110d3

**Comment by greg-schroeder** [comment]
Awaiting review from @aimane-chnaif

**Comment by greg-schroeder** [comment]
Bump @aimane-chnaif

**Comment by zaitsev1** [comment]
@tienifr Could you share a video of how it works for you?

if you want to access the `inputRef` of a component, you cannot do it directly in the `componentDidMount` lifecycle method of the `ScreenWrapper` component. The reason is that the `inputRef` is not yet available in the component's DOM at the time when `componentDidMount` is called.

To access the `inputRef`, you need to use a listener event, such as the `transitionEnd` event. However, but the `transitionEnd` event is specific to React Navigation and is triggered only when navigating between screens using the navigation methods provided by React Navigation, such as navigate, push, goBack, etc.

https://github.com/Expensify/App/assets/32131367/d8ab023f-4d7d-497a-9de3-6cde8a23be74

**Comment by zaitsev1** [comment]
My [proposal](https://github.com/Expensify/App/issues/22850#issuecomment-1642188069) has been updated

**Comment by greg-schroeder** [comment]
bump @aimane-chnaif to check updated proposals

**Comment by greg-schroeder** [comment]
bump on the above!

**Comment by greg-schroeder** [comment]
Okay waiting on @aimane-chnaif's proposal review!

**Comment by greg-schroeder** [comment]
Same as above

**Comment by greg-schroeder** [comment]
I can place it on hold if the solution depends on that. Shall we re-open that issue you linked? It's closed at the moment

**Comment by greg-schroeder** [comment]
Sounds good. Reopened that and held this one.

**Comment by hannojg** [comment]
> The Onyx update is to use cached data if available

Yes, its on main now!

[Not sure if that's what you wanted me to confirm, I haven't read through all messages here 🙇 ]

**Comment by b4s36t4** [comment]
@bernhardoj this merely reproduce-able on prod, It surely re-produceable on dev though.

**Comment by greg-schroeder** [comment]
Hmm - okay, so should this remain on hold, or should we close it and take care of it in #22508, or create a new issue entirely?

**Comment by greg-schroeder** [comment]
Okay, got it, removed hold

**Comment by greg-schroeder** [comment]
Seems we're awaiting a viable proposal here now

**Comment by greg-schroeder** [comment]
Bump @aimane-chnaif

**Comment by greg-schroeder** [comment]
bump @aimane-chnaif for proposals

**Comment by zaitsev1** [comment]
This [proposal](https://github.com/Expensify/App/issues/22850#issuecomment-1719338883) does not work for me:

https://github.com/Expensify/App/assets/32131367/0c825de9-cd2b-4884-8d4f-6f1830a40b49

**I believe that the best way to implement focusing would be through a Form component that wraps all inputs**

**Comment by b4s36t4** [comment]
@aimane-chnaif I don't think using `useFocusEffect` and `setTimeout` can solve this issue fully maybe we get new issue related to this again. 

Upon my investigation with `useFocusEffect`, it uses the `navigation.addEventListener` to subscribe to `focus` event from `react-navigation`. As describe on my previous RCA why `onEntryTransitionEnd` not firing we can create a new hook to catch the `transitionEnd` events and run the callback based on it?

maybe like following

```jsx
const useFocusTransition = (callback) => {
    const navigation = useNavigation()
    useEffect(() => {
        navigation.addEventListener('focus', callback)
        return () => navigation.removeListener('focus', callback)
    },[navigation])
}
```

Tried on a component as far as result is promising. Any thoughts on this proposal?

**Comment by zaitsev1** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Input field is not focused after reloading.

### What is the root cause of that problem?
The [hook](https://github.com/Expensify/App/issues/22850#issuecomment-1719338883) doesn't guarantee that `inputRef` has been initialized when attempting to focus on it. The code uses a timeout to attempt the focus, but it doesn't ensure that the input element is available at that moment.

### What changes do you think we should make in order to solve the problem?

updated hook:

```javascript
const useAutoFocusInput = () => {
    const [isInputInitialized, setIsInputInitialized] = useState(false);

    const inputRef = useRef(null);
    const focusTimeoutRef = useRef(null);

    useFocusEffect(
        useCallback(() => {
            focusTimeoutRef.current = setTimeout(() => inputRef.current && inputRef.current.focus(), CONST.ANIMATED_TRANSITION);
            return () => {
                if (!focusTimeoutRef.current) {
                    return;
                }
                clearTimeout(focusTimeoutRef.current);
            };
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [isInputInitialized]),
    );

    const inputCallbackRef = (ref) => {
        inputRef.current = ref;
        setIsInputInitialized(true);
    };

    return {inputRef, inputCallbackRef};
};
``` 

The updated code ensures that the focus attempt occurs only after the input element has been properly initialized and is ready to receive focus.

You can use `useAutoFocusInput` like this:

```javascript
const { inputRef, inputCallbackRef } = useAutoFocusInput();

return (
  <TextInput
    ref={inputCallbackRef}
    // ... other TextInput props
  />
);
```

Video Demo: 

[Video Link](https://github.com/Expensify/App/assets/32131367/e377a759-824c-4e86-8b1d-9f5788246de0)

### What alternative solutions did you explore? (Optional)
NA

**Comment by zaitsev1** [comment]
@aimane-chnaif PR is ready to review https://github.com/Expensify/App/pull/27982 🚀

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by stitesExpensify** [comment]
To confirm, is this still reproducible? It looks like Applause couldn't reproduce this week?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by greg-schroeder** [comment]
Processing

**Comment by greg-schroeder** [comment]
Issue Participants:

Issue reported by: @gadhiyamanan
Contributor: @tienifr 
C+: @aimane-chnaif

Was this issue merged in time to be eligible for the speed bonus? y
Were their any regressions? n

Payment summary:

Reporter: $250
Contributor: $1,500
C+: $1,500

**Comment by greg-schroeder** [comment]
Offers sent. Next up is @aimane-chnaif to complete the c+ checklist

**Comment by greg-schroeder** [comment]
Thanks! Completing payments and closing.
